test/Github/Tests/Functional/CacheTest.php

```diff
@@ -0,0 +1,69 @@
+<?php
+
+namespace Github\Tests\Functional;
+
+use Cache\Adapter\PHPArray\ArrayCachePool;
+use Github\Client;
+use GuzzleHttp\Psr7\Response;
+use function GuzzleHttp\Psr7\stream_for;
+
+/**
+ * @group functional
+ *
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ */
+class CacheTest extends \PHPUnit_Framework_TestCase
+{
+    /**
+     * @test
+     */
+    public function shouldServeCachedResponse()
+    {
+        $mockClient = new \Http\Mock\Client();
+        $mockClient->addResponse($this->getCurrentUserResponse('nyholm'));
+        $mockClient->addResponse($this->getCurrentUserResponse('octocat'));
+
+        $github = Client::createWithHttpClient($mockClient);
+        $github->addCache(new ArrayCachePool(), ['default_ttl'=>600]);
+
+        $github->authenticate('fake_token_aaa', Client::AUTH_HTTP_TOKEN);
+        $userA = $github->currentUser()->show();
+        $this->assertEquals('nyholm', $userA['login']);
+
+        $userB = $github->currentUser()->show();
+        $this->assertEquals('nyholm', $userB['login'], 'Two request following each other should be cached.');
```

Comment:
> Client should always send a request, and then see if we get a "not modified" response

No, not always. If server says that it is okey to cache for X time, we do not need to send a request for that period. 
But the logic you are referring to is in the php-http/cache-plugin. 

> We shouldn't just randomly cache for 600 seconds.
I set the default ttl to arbitrary number here to force the first the request to be cached (for a time longer then the tests run). Doing that I can easily verify if we make a second request or not. (Im using the MockClient, Im not actually making any HTTP calls. )